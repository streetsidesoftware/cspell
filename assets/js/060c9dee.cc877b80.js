"use strict";(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[94023],{46039:(e,t,n)=>{n.d(t,{R:()=>l,x:()=>c});var i=n(18225);const s={},r=i.createContext(s);function l(e){const t=i.useContext(r);return i.useMemo((function(){return"function"==typeof e?e(t):{...t,...e}}),[t,e])}function c(e){let t;return t=e.disableParentContext?"function"==typeof e.components?e.components(s):e.components||s:l(e.components),i.createElement(r.Provider,{value:t},e.children)}},87257:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>a,contentTitle:()=>c,default:()=>p,frontMatter:()=>l,metadata:()=>i,toc:()=>o});const i=JSON.parse('{"id":"api/cspell-lib/functions/readRawSettings","title":"Function: readRawSettings()","description":"readRawSettings(filename, relativeTo?): Promise\\\\","source":"@site/docs/api/cspell-lib/functions/readRawSettings.md","sourceDirName":"api/cspell-lib/functions","slug":"/api/cspell-lib/functions/readRawSettings","permalink":"/docs/api/cspell-lib/functions/readRawSettings","draft":false,"unlisted":false,"editUrl":"https://github.com/streetsidesoftware/cspell/tree/main/website/docs/docs/api/cspell-lib/functions/readRawSettings.md","tags":[],"version":"current","frontMatter":{},"sidebar":"tutorialSidebar","previous":{"title":"Function: readFileSync()","permalink":"/docs/api/cspell-lib/functions/readFileSync"},"next":{"title":"Function: readSettings()","permalink":"/docs/api/cspell-lib/functions/readSettings"}}');var s=n(37557),r=n(46039);const l={},c="Function: readRawSettings()",a={},o=[{value:"Parameters",id:"parameters",level:2},{value:"filename",id:"filename",level:3},{value:"relativeTo?",id:"relativeto",level:3},{value:"Returns",id:"returns",level:2}];function d(e){const t={a:"a",blockquote:"blockquote",code:"code",h1:"h1",h2:"h2",h3:"h3",header:"header",p:"p",strong:"strong",...(0,r.R)(),...e.components};return(0,s.jsxs)(s.Fragment,{children:[(0,s.jsx)(t.header,{children:(0,s.jsx)(t.h1,{id:"function-readrawsettings",children:"Function: readRawSettings()"})}),"\n",(0,s.jsxs)(t.blockquote,{children:["\n",(0,s.jsxs)(t.p,{children:[(0,s.jsx)(t.strong,{children:"readRawSettings"}),"(",(0,s.jsx)(t.code,{children:"filename"}),", ",(0,s.jsx)(t.code,{children:"relativeTo"}),"?): ",(0,s.jsx)(t.code,{children:"Promise"}),"<",(0,s.jsx)(t.a,{href:"/docs/api/cspell-lib/interfaces/CSpellSettingsWithSourceTrace",children:(0,s.jsx)(t.code,{children:"CSpellSettingsWithSourceTrace"})}),">"]}),"\n"]}),"\n",(0,s.jsxs)(t.p,{children:["Defined in: ",(0,s.jsx)(t.a,{href:"https://github.com/streetsidesoftware/cspell/blob/3fc3ef2ea9def847ef20287c6238ff423e6493eb/packages/cspell-lib/src/lib/Settings/Controller/configLoader/defaultConfigLoader.ts#L99",children:"packages/cspell-lib/src/lib/Settings/Controller/configLoader/defaultConfigLoader.ts:99"})]}),"\n",(0,s.jsx)(t.h2,{id:"parameters",children:"Parameters"}),"\n",(0,s.jsx)(t.h3,{id:"filename",children:"filename"}),"\n",(0,s.jsxs)(t.p,{children:[(0,s.jsx)(t.code,{children:"string"})," | ",(0,s.jsx)(t.code,{children:"URL"})]}),"\n",(0,s.jsx)(t.h3,{id:"relativeto",children:"relativeTo?"}),"\n",(0,s.jsxs)(t.p,{children:[(0,s.jsx)(t.code,{children:"string"})," | ",(0,s.jsx)(t.code,{children:"URL"})]}),"\n",(0,s.jsx)(t.h2,{id:"returns",children:"Returns"}),"\n",(0,s.jsxs)(t.p,{children:[(0,s.jsx)(t.code,{children:"Promise"}),"<",(0,s.jsx)(t.a,{href:"/docs/api/cspell-lib/interfaces/CSpellSettingsWithSourceTrace",children:(0,s.jsx)(t.code,{children:"CSpellSettingsWithSourceTrace"})}),">"]})]})}function p(e={}){const{wrapper:t}={...(0,r.R)(),...e.components};return t?(0,s.jsx)(t,{...e,children:(0,s.jsx)(d,{...e})}):d(e)}}}]);