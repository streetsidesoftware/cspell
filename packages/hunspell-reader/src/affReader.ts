import assert from 'assert';
import { readFile } from 'fs/promises';
import { decode } from 'iconv-lite';

import { Aff } from './aff';
import type { AffInfo, Fx, Rep, SubstitutionSet } from './affDef';
import { cleanObject, isDefined } from './util';

const fixRegex = {
    SFX: { m: /$/, r: '$' },
    PFX: { m: /^/, r: '^' },
};

const yesRegex = /[yY]/;
const spaceRegex = /\s+/;
const commentRegex = /(?:^\s*#.*)|(?:\s+#.*)/;
const affixLine = /^\s*([^\s]+)\s+(.*)?$/;

const UTF8 = 'UTF-8';

interface Collector<T> {
    addLine(line: AffLine): void;
    getValue(): T | undefined;
}

export interface ConvEntry {
    from: string;
    to: string;
}

function convEntry(): Collector<ConvEntry[]> {
    let fieldValue: ConvEntry[] | undefined;

    return {
        addLine: (line: AffLine) => {
            if (fieldValue === undefined) {
                fieldValue = [];
                return;
            }
            const args = (line.value || '').split(spaceRegex);
            fieldValue.push({ from: args[0], to: args[1] });
        },
        getValue: () => fieldValue,
    };
}

function afEntry() {
    let fieldValue: string[] | undefined;

    return {
        addLine: (line: AffLine) => {
            if (fieldValue === undefined) {
                // Add empty entry because rules start at 1
                fieldValue = [''];
                return;
            }
            if (line.value) {
                fieldValue.push(line.value);
            }
        },
        getValue: () => fieldValue,
    };
}

function simpleTable<T>(map: (values: string[][]) => T) {
    let data:
        | {
              count: string;
              extra: string[] | undefined;
              values: string[][];
          }
        | undefined;

    function getValue() {
        if (data?.values) return map(data.values);
        return undefined;
    }

    function addLine(line: AffLine): void {
        const args = (line.value || '').split(spaceRegex);
        if (data === undefined) {
            const [count, ...extraValues] = args;
            const extra = extraValues.length ? extraValues : undefined;
            const values: string[][] = [];
            data = { count, extra, values };
            return;
        }

        data.values.push(args);
    }

    return { addLine, getValue };
}

function tablePfxOrSfx(fieldValue: Afx | undefined, line: AffLine): Afx {
    /*
    Fields of an affix rules:
    (0) Option name
    (1) Flag
    (2) stripping characters from beginning (at prefix rules) or end (at suffix rules) of the word
    (3) affix (optionally with flags of continuation classes, separated by a slash)
    (4) condition.
        Zero stripping or affix are indicated by zero. Zero condition is indicated by dot.
        Condition is a simplified, regular expression-like pattern, which must be met before the affix can be applied.
        (Dot signs an arbitrary character. Characters in braces sign an arbitrary character from the character subset.
        Dash hasn't got special meaning, but circumflex (^) next the first brace sets the complimenter character set.)
    (5) Optional morphological fields separated by spaces or tabulators.
     */
    if (fieldValue === undefined) {
        fieldValue = new Map<string, Fx>();
    }
    const [subField] = (line.value || '').split(spaceRegex);
    if (!fieldValue.has(subField)) {
        const fx = parseAffixCreation(line);
        fieldValue.set(fx.id, fx);
        return fieldValue;
    }
    const rule = parseAffixRule(line);
    if (!rule) {
        console.log(`Affix rule missing values: ${line.option} ${line.value}`);
        return fieldValue;
    }

    const fixRuleSet = fieldValue.get(subField);
    assert(fixRuleSet);
    const substitutionSets = fixRuleSet.substitutionSets;
    const ruleAsString = rule.condition.source;
    if (!substitutionSets.has(ruleAsString)) {
        substitutionSets.set(ruleAsString, {
            match: rule.condition,
            substitutions: [],
        });
    }
    const substitutionSet = substitutionSets.get(ruleAsString);
    assert(substitutionSet);
    const [attachText, attachRules] = rule.affix.split('/', 2);
    substitutionSet.substitutions.push({
        remove: rule.stripping,
        replace: rule.replace,
        attach: attachText,
        attachRules,
        extra: rule.extra,
    });

    return fieldValue;
}

/**
 * Parse Affix creation line:
 * `PFX|SFX flag cross_product number`
 */
function parseAffixCreation(line: AffLine): Fx {
    const [flag, combinable, count, ...extra] = (line.value || '').split(spaceRegex);
    const fx: Fx = {
        id: flag,
        type: line.option,
        combinable: !!combinable.match(yesRegex),
        count,
        extra,
        substitutionSets: new Map<string, SubstitutionSet>(),
    };
    return fx;
}

interface AffixRule {
    type: 'PFX' | 'SFX';
    flag: string;
    stripping: string;
    replace: RegExp;
    affix: string;
    condition: RegExp;
    extra?: string;
}

const affixRuleRegEx = /^(\S+)\s+(\S+)\s+(\S+)\s*(.*)/;
const affixRuleConditionRegEx = /^((?:\[.*\]|\S+)+)\s*(.*)/;

/**
 * `PFX|SFX flag stripping prefix [condition [morphological_fields...]]`
 */
function parseAffixRule(line: AffLine): AffixRule | undefined {
    const [, flag, strip, affix, optional = ''] = (line.value || '').match(affixRuleRegEx) || [];
    if (!flag || !strip || !affix) {
        return undefined;
    }
    const [, rawCondition = '.', extra] = optional.match(affixRuleConditionRegEx) || [];

    const type = line.option === 'SFX' ? 'SFX' : 'PFX';
    const condition = fixMatch(type, rawCondition);

    const affixRule: AffixRule = {
        type,
        flag,
        stripping: strip,
        replace: fixMatch(type, strip),
        affix: cleanAffixAttach(affix),
        condition,
        extra,
    };
    return affixRule;
}

function cleanAffixAttach(affix: string): string {
    const [fix, rules] = affix.split('/', 2);

    const attach = fix === '0' ? '' : fix;
    return attach + (rules ? '/' + rules : '');
}

function fixMatch(type: AffixRule['type'], match: string): RegExp {
    const exp = affixMatchToRegExpString(match);
    const fix = fixRegex[type];
    return new RegExp(exp.replace(fix.m, fix.r));
}

function affixMatchToRegExpString(match: string): string {
    if (match === '0') return '';
    return match.replace(/([\\\-?*])/g, '\\$1');
}

function collectFx(): Collector<Afx> {
    let value: Afx | undefined;
    function addLine(line: AffLine) {
        value = tablePfxOrSfx(value, line);
    }
    return {
        addLine,
        getValue: () => value,
    };
}

const asPfx = collectFx;
const asSfx = collectFx;

const asString = () => collectPrimitive<string>((v) => v, '');
const asBoolean = () => collectPrimitive<boolean>((v) => !!parseInt(v), '1');
const asNumber = () => collectPrimitive<number>(parseInt, '0');

function collectPrimitive<T>(map: (line: string) => T, defaultValue = ''): Collector<T> {
    let primitive: T | undefined;

    function getValue() {
        return primitive;
    }

    function addLine(line: AffLine) {
        const { value = defaultValue } = line;
        primitive = map(value);
    }

    return { addLine, getValue };
}

function toRep(values: string[][]): Rep[] {
    return values.map((v) => ({ match: v[0], replaceWith: v[1] }));
}

function toSingleStrings(values: string[][]): string[] {
    return values.map((v) => v[0]).filter(isDefined);
}

function toAffMap(values: string[][]): Exclude<AffInfo['MAP'], undefined> {
    return toSingleStrings(values);
}

function toCompoundRule(values: string[][]): Exclude<AffInfo['COMPOUNDRULE'], undefined> {
    return toSingleStrings(values);
}

function toCheckCompoundPattern(values: string[][]): Exclude<AffInfo['CHECKCOMPOUNDPATTERN'], undefined> {
    return values;
}

type FieldCollector<T> = Collector<T>;

type AffFieldCollectorTable = {
    [key in keyof AffInfo]-?: FieldCollector<Exclude<AffInfo[key], undefined>>;
};

/*
cspell:ignore COMPOUNDBEGIN COMPOUNDEND COMPOUNDMIDDLE COMPOUNDMIN COMPOUNDPERMITFLAG COMPOUNDRULE COMPOUNDFORBIDFLAG COMPOUNDFLAG
cspell:ignore FORBIDDENWORD KEEPCASE
cspell:ignore MAXDIFF NEEDAFFIX WORDCHARS
*/

// prettier-ignore
const createAffFieldTable: () => AffFieldCollectorTable = () => ({
    AF                  : afEntry(),
    BREAK               : simpleTable(toSingleStrings),
    CHECKCOMPOUNDCASE   : asBoolean(),
    CHECKCOMPOUNDDUP    : asBoolean(),
    CHECKCOMPOUNDPATTERN: simpleTable(toCheckCompoundPattern),
    CHECKCOMPOUNDREP    : asBoolean(),
    COMPOUNDBEGIN       : asString(),
    COMPOUNDEND         : asString(),
    COMPOUNDMIDDLE      : asString(),
    COMPOUNDMIN         : asNumber(),
    COMPOUNDFLAG        : asString(),
    COMPOUNDPERMITFLAG  : asString(),
    COMPOUNDFORBIDFLAG  : asString(),
    COMPOUNDRULE        : simpleTable(toCompoundRule),
    FLAG                : asString(), // 'long' | 'num'
    FORBIDDENWORD       : asString(),
    FORCEUCASE          : asString(),
    ICONV               : convEntry(),
    KEEPCASE            : asString(),
    KEY                 : asString(),
    MAP                 : simpleTable(toAffMap),
    MAXCPDSUGS          : asNumber(),
    MAXDIFF             : asNumber(),
    NEEDAFFIX           : asString(),
    NOSPLITSUGS         : asBoolean(),
    NOSUGGEST           : asString(),
    OCONV               : convEntry(),
    ONLYINCOMPOUND      : asString(),
    ONLYMAXDIFF         : asBoolean(),
    PFX                 : asPfx(),
    REP                 : simpleTable(toRep),
    SET                 : asString(),
    SFX                 : asSfx(),
    TRY                 : asString(),
    WARN                : asString(),
    WORDCHARS           : asString(),
});

function collectionToAffInfo(affFieldCollectionTable: AffFieldCollectorTable, encoding: string): AffInfo {
    type AffInfoKeys = keyof AffInfo;
    type ParseResult = {
        [key in AffInfoKeys]: AffInfo[key];
    };

    // prettier-ignore
    const result: ParseResult = {
        AF                  : affFieldCollectionTable.AF.getValue(),
        BREAK               : affFieldCollectionTable.BREAK.getValue(),
        CHECKCOMPOUNDCASE   : affFieldCollectionTable.CHECKCOMPOUNDCASE.getValue(),
        CHECKCOMPOUNDDUP    : affFieldCollectionTable.CHECKCOMPOUNDDUP.getValue(),
        CHECKCOMPOUNDPATTERN: affFieldCollectionTable.CHECKCOMPOUNDPATTERN.getValue(),
        CHECKCOMPOUNDREP    : affFieldCollectionTable.CHECKCOMPOUNDREP.getValue(),
        COMPOUNDBEGIN       : affFieldCollectionTable.COMPOUNDBEGIN.getValue(),
        COMPOUNDEND         : affFieldCollectionTable.COMPOUNDEND.getValue(),
        COMPOUNDMIDDLE      : affFieldCollectionTable.COMPOUNDMIDDLE.getValue(),
        COMPOUNDMIN         : affFieldCollectionTable.COMPOUNDMIN.getValue(),
        COMPOUNDFLAG        : affFieldCollectionTable.COMPOUNDFLAG.getValue(),
        COMPOUNDPERMITFLAG  : affFieldCollectionTable.COMPOUNDPERMITFLAG.getValue(),
        COMPOUNDFORBIDFLAG  : affFieldCollectionTable.COMPOUNDFORBIDFLAG.getValue(),
        COMPOUNDRULE        : affFieldCollectionTable.COMPOUNDRULE.getValue(),
        FLAG                : affFieldCollectionTable.FLAG.getValue(),
        FORBIDDENWORD       : affFieldCollectionTable.FORBIDDENWORD.getValue(),
        FORCEUCASE          : affFieldCollectionTable.FORCEUCASE.getValue(),
        ICONV               : affFieldCollectionTable.ICONV.getValue(),
        KEEPCASE            : affFieldCollectionTable.KEEPCASE.getValue(),
        KEY                 : affFieldCollectionTable.KEY.getValue(),
        MAP                 : affFieldCollectionTable.MAP.getValue(),
        MAXCPDSUGS          : affFieldCollectionTable.MAXCPDSUGS.getValue(),
        MAXDIFF             : affFieldCollectionTable.MAXDIFF.getValue(),
        NEEDAFFIX           : affFieldCollectionTable.NEEDAFFIX.getValue(),
        NOSPLITSUGS         : affFieldCollectionTable.NOSPLITSUGS.getValue(),
        NOSUGGEST           : affFieldCollectionTable.NOSUGGEST.getValue(),
        OCONV               : affFieldCollectionTable.OCONV.getValue(),
        ONLYINCOMPOUND      : affFieldCollectionTable.ONLYINCOMPOUND.getValue(),
        ONLYMAXDIFF         : affFieldCollectionTable.ONLYMAXDIFF.getValue(),
        PFX                 : affFieldCollectionTable.PFX.getValue(),
        REP                 : affFieldCollectionTable.REP.getValue(),
        SET                 : affFieldCollectionTable.SET.getValue() || encoding,
        SFX                 : affFieldCollectionTable.SFX.getValue(),
        TRY                 : affFieldCollectionTable.TRY.getValue(),
        WARN                : affFieldCollectionTable.WARN.getValue(),
        WORDCHARS           : affFieldCollectionTable.WORDCHARS.getValue(),
    };
    return cleanObject(result);
}

export async function parseAffFile(filename: string, encoding: string = UTF8) {
    const buffer = await readFile(filename);
    const file = decode(buffer, encoding);
    const affInfo = parseAff(file, encoding);
    if (affInfo.SET && affInfo.SET.toLowerCase() !== encoding.toLowerCase()) {
        return parseAff(decode(buffer, affInfo.SET.toLowerCase()), affInfo.SET);
    }
    return affInfo;
}

export function parseAff(affFileContent: string, encoding: string = UTF8): AffInfo {
    const lines = affFileContent.split(/\r?\n/g);
    const affFieldCollectionTable = createAffFieldTable();
    affFieldCollectionTable.SET.addLine({ option: 'SET', value: encoding });
    lines
        .map((line) => line.trimStart())
        .map((line) => line.replace(commentRegex, ''))
        .filter((line) => line.trim() !== '')
        .map(parseLine)
        .forEach((line: AffLine) => {
            const field = line.option as keyof AffInfo;
            affFieldCollectionTable[field]?.addLine(line);
        });
    return collectionToAffInfo(affFieldCollectionTable, encoding);
}

export function parseAffFileToAff(filename: string, encoding?: string): Promise<Aff> {
    return parseAffFile(filename, encoding).then((affInfo) => new Aff(affInfo));
}

function parseLine(line: string): AffLine {
    const result = line.match(affixLine) || ['', ''];
    const [, option, value] = result;
    return { option, value: value || undefined };
}

export interface AffLine {
    option: string;
    value: string | undefined;
}

type Afx = Map<string, Fx>;

export const testing = {
    parseAffixRule,
    tablePfxOrSfx,
    parseLine,
};
