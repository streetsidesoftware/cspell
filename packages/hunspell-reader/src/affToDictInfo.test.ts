import path from 'path';
import { describe, expect, test } from 'vitest';

import { parseAffFile } from './affReader.js';
import { affToDicInfo } from './affToDicInfo.js';

const dictionaries = path.join(__dirname, '../dictionaries');
const sampleAff = path.join(dictionaries, 'Portuguese (Brazilian).aff');
const sampleLocale = 'pt-br';

describe('affToDictInfo', () => {
    test('parseAffFile', async () => {
        const affInfo = await parseAffFile(sampleAff);
        const info = affToDicInfo(affInfo, sampleLocale);
        expect(info).toMatchSnapshot();
    });
});
