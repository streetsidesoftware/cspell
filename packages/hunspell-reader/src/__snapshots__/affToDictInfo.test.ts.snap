// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`affToDictInfo > parseAffFile 1`] = `
{
  "accents": "̀-̧̃̈",
  "alphabet": "A-Za-zÀÁÂÃÇÉÊÍÓÔÕÚÜàáâãçéêíóôõúü",
  "locale": "pt-br",
  "suggestionEditCosts": [
    {
      "description": "Capitalization change.",
      "map": "Aa|Bb|Cc|Dd|Ee|Ff",
      "replace": 1,
    },
    {
      "description": "Capitalization change.",
      "map": "Gg|Hh|Ii|Jj|Kk|Ll",
      "replace": 1,
    },
    {
      "description": "Capitalization change.",
      "map": "Mm|Nn|Oo|Pp|Qq|Rr",
      "replace": 1,
    },
    {
      "description": "Capitalization change.",
      "map": "Ss|Tt|Uu|Vv|Ww|Xx",
      "replace": 1,
    },
    {
      "description": "Capitalization change.",
      "map": "Yy|Zz|Àà|Áá|Ââ|Ãã",
      "replace": 1,
    },
    {
      "description": "Capitalization change.",
      "map": "Çç|Éé|Êê|Íí|Óó|Ôô",
      "replace": 1,
    },
    {
      "description": "Capitalization change.",
      "map": "Õõ|Úú|Üü",
      "replace": 1,
    },
    {
      "description": "Replace Accents",
      "map": "AÀÁÂÃ|CÇ|EÉÊ|IÍ|OÓÔÕ|UÚÜ",
      "replace": 1,
    },
    {
      "description": "Replace Accents",
      "map": "aàáâã|cç|eéê|ií|oóôõ|uúü",
      "replace": 1,
    },
    {
      "description": "Capitalization and Accent change.",
      "map": "AaÀÁÂÃàáâã|CcÇç|EeÉÊéê|IiÍí|OoÓÔÕóôõ|UuÚÜúü",
      "replace": 2,
    },
    {
      "description": "Hunspell Aff Map",
      "map": "aáãâAÁÃÂ|cçCÇ|eéêEÉÊ|iíIÍ|oóõôOÓÕÔ|uúüUÚÜ",
      "replace": 1,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(-a-)(_a_)|(-chuvas)(s-chuvas)|(-com-)(_com_)|(-da-)(_da_)|(-de-)(_de_)|(-l)l",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(-la-)(la)|(-las-)(las)|(-lhe-)(lhe)|(-lhes-)(lhes)|(-lo-)(lo)|(-los-)(los)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(-me-)(me)|(-nos-)(nos)|(-papéis)(s-papel)|(-se-)(se)|(-te-)(te)|(-vos-)(vos)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(Cambojano)(Khmer)|(Cuanhama)(Kwanyama)|(Kuwait)(Quait)|(acc)(hak)|(ackup)(ecape)|(ad)(izad)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(ada)(ável)|(adas)(áveis)|(ado)(ável)|(ados)(áveis)|(agu)(águ)|(ainers)(êineres)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(ala)(á-la)|(ala)(á-las)|(alh)(áli)|(all)(ol)|(alo)(á-lo)|(alos)(á-los)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(am)(an)|(am)(ão)|(am)ã|(an)ã|(ao)(ão)|(aqu)(áqu)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(ara)(ára-)|(ars)(raç)|(athlon)(átlon)|(audio)(áudio-)|(ayout)(eiaute)|(az)(ás)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(azinh)(inh)|(bakup)(becape)|(bilidade)(lidade)|(boy)(bói)|(bstâ)(bta)|(bstâ)(btâ)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(ca)(ka)|(cao)(ção)|(ce)(sse)|(celofa)(selofo)|(ch)(sh)|(ch)x",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(che)(sh)|(chi)(qui)|(ci)(ssi)|(cie)(sciê)|(cinss)(sinc)|(ck)(que)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(co)(ko)|(cs)x|(cu)(ku)|(cá)(ká)|(cão)(ção)|(cç)x",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(cõ)(çõ)|(defissio)(difícil)|(des)(dis)|(difissi)(difíci)|(difissio)(difícil)|(e_)é",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(ebam)(ibã)|(eban)(ibã)|(ebã)(ibã)|(eem)(êm)|(ei)(éi)|(el)(éis)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(elh)(éli)|(em)(en)|(empecilh)(impecíli)|(ensi)(ênci)|(env)v|(eoro)(ereo)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(ept)(épit)|(erug)(irurg)|(es)(ex)|(es)(ez)|(escân)(scann)|(escâneres)(scaners)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(escâneres)(scanners)|(esqueite)(skat)|(esqueiti)(skati)|(estú)(stu)|(ez)(és)|(ez)(ês)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(ezinh)(inh)|(ffice)(ffice.org)|(gcoc)(guecoque)|(ge)(je)|(gem)(mento)|(gen)(mento)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(gi)(ji)|(gi)j|(gkok)(guecoque)|(gs)(gues)|(gê)(jê)|(gí)(jí)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(hkek)(queque)|(ida)(ível)|(idas)(íveis)|(ido)(ível)|(idos)(íveis)|(ifem)(ífen)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(ilh)(íli)|(ilzinh)(inh)|(im)(in)|(in)(nhe)|(in)(nhi)|(in)(não-)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(in)(não_)|(inh)(ozinh)|(ionguiangue)(yongyang)|(iquiavique)(ykjavík)|(ir)(não-)|(ji)(jing)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(ke)(que)|(ki)(qui)|(kiwi)(quiuí)|(kê)(quê)|(l_G)(lg)|(leb)(reb)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(lh)(li)|(ls)(lç)|(ls)(uç)|(lç)(uç)|(m_)n|(mento)(ção)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(mento)(çõe)|(o-eletrô)(oeletro)|(o_)ó|(obr)(robl)|(ografista)(ógrafo)|(oi)(ói)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(oky)(óqui)|(olh)(óli)|(ologista)(ólogo)|(om)(on)|(oss)(ós-s)|(paraq)(pára-q)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(perco)(perdo)|(ping)(pingue-pongue)|(ping-pong)(pingue-pongue)|(pingue-pongue)(pong)|(pre)(pré-)|(qs)(ques)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(quinque)(quinquê)|(quinque)(quinqüe)|(quinque)(qüinque)|(quinqui)(quinquí)|(quinqui)(quinqüi)|(quinqui)(qüinqui)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(quinquê)(quinquê)|(quinquê)(quinqüe)|(quinquê)(quinqüê)|(quinquê)(qüinque)|(quinquê)(qüinquê)|(quinquí)(quinquí)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(quinquí)(quinqüi)|(quinquí)(quinqüí)|(quinquí)(qüinqui)|(quinquí)(qüinquí)|(r-de)(res-de)|(r-lhe-ia)(ria-lhe)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(r-lhe-iam)(riam-lhe)|(r-lhes-ia)(ria-lhes)|(r-lhes-iam)(riam-lhes)|(r-me-ia)(ria-me)|(r-me-iam)(riam-me)|(r-nos-ia)(ria-nos)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(r-nos-iam)(riam-nos)|(r-se-ia)(ria-se)|(r-te-ia)(ria-te)|(r-te-iam)(riam-te)|(r-vos-ia)(ria-vos)|(r-vos-iam)(riam-vos)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(radio)(rádio-)|(revigoração)(revitalização)|(ridade)(ridão)|(ridade)(ridõe)|(rs)(ss)|(rsa)(rça)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(sciê)(sie)|(septec)(septing)|(septic)(septing)|(septing)(setec)|(socio)(sócio-)|(ss)ç",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(ta)(tiva)|(tativ)(tiv)|(tic-tac)(tique-taque)|(tividade)(tivide)|(uait)(uwait)|(ulh)(úli)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(um)(un)|(uê)(üe)|(uê)(üê)|(video)(vídeo-)|(video)(vídeos-)|(vigoração)(vigência)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(zavil)(zzavill)|(âm)(ân)|(ãe)(ão)|(ãe)(õe)|(ão)(õe)|(ém)(én)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "(êm)(ên)|(ím)(ín)|(óm)(ón)|(ôm)(ôn)|(úm)(ún)|-(s-)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "-_|-r|-s|a(a-)|a(a-h)|a(a_)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "aá|aâ|aã|b(b-)|c(cc)|c(qu)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "c(sc)|cs|cç|e(e-)|e(e-h)|e(e_)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "e(é-)|eé|eê|f(f.)|f(ph)|g(gu)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "g(gue)|gj|i(i-)|i(i-h)|i(i_)|ií",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "l(l-)|l(l_)|lu|m(m-)|m(m_)|n(n-)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "o(o-)|o(o-h)|o(o_)|o(ó-)|oó|oô",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "oõ|q(que)|r(r-)|r(r_)|r(rr)|s(s-)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "s(s.)|s(s_)|sz|sç|u(u-)|u(u_)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "uú|v(v.)|x(x.)|x(x_)|z(z-)|z(z_)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "á(á-)|á(á_)|âã|ã(ã-)|ã(ã_)|é(é-)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "é(é-h)|é(é_)|éê|ê(ê-)|ê(ê_)|í(í-)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "í(í_)|ó(ó-)|ó(ó-h)|ó(ó_)|óô|ô(ô-)",
      "replace": 75,
    },
    {
      "description": "Hunspell Replace Map",
      "map": "ô(ô_)|ú(ú_)",
      "replace": 75,
    },
  ],
}
`;
